src/ripple/protocol/LedgerEntryWrapper.h

```diff
@@ -0,0 +1,194 @@
+    // There is no virtual table, so the most derived class must be destroyed.
+    ~LedgerEntryWrapper() = default;
+
+    //--------------------------------------------------------------------------
+    // Helper functions that are useful to some derived classes.
+    template <typename SF, typename T>
+    void
+    setOptional(SF const& field, T const& value)
+    {
+        static_assert(Writable, "Cannot set member of const ledger entry.");
+        static_assert(
+            std::is_base_of_v<SField, SF>,
+            "setOptional()requires an SField as its first argument.");
+
+        if (!wrapped_->isFieldPresent(field))
+            wrapped_->makeFieldPresent(field);
+
+        wrapped_->at(field) = value;
+    }
+
+    template <typename SF>
+    void
+    clearOptional(SF const& field)
+    {
+        static_assert(Writable, "Cannot set member of const ledger entry.");
+        static_assert(
+            std::is_base_of_v<SField, SF>,
+            "clearOptional()requires an SField as its argument.");
+
+        if (wrapped_->isFieldPresent(field))
+            wrapped_->makeFieldAbsent(field);
+    }
+
+    [[nodiscard]] Blob
+    getOptionalVL(SF_VL const& field) const
+    {
+        Blob ret;
+        if (wrapped_->isFieldPresent(field))
+            ret = wrapped_->getFieldVL(field);
+        return ret;
+    }
+
+    template <typename SF, std::size_t Bits, class Tag>
+    void
+    setOrClearBaseUintIfZero(SF const& field, base_uint<Bits, Tag> const& value)
+    {
+        static_assert(Writable, "Cannot set member of const ledger entry.");
+        static_assert(
+            std::is_base_of_v<SField, SF>,
+            "setOrClearBaseUintIfZero()requires an SField as its argument.");
+
+        if (value.signum() == 0)
+            return clearOptional(field);
+
+        if (!wrapped_->isFieldPresent(field))
+            wrapped_->makeFieldPresent(field);
+
+        wrapped_->at(field) = value;
+    }
+
+    void
+    setOrClearVLIfEmpty(SF_VL const& field, Blob const& value)
+    {
+        static_assert(Writable, "Cannot set member of const ledger entry.");
+
+        if (value.empty())
+            return clearOptional(field);
+
+        if (!wrapped_->isFieldPresent(field))
+            wrapped_->makeFieldPresent(field);
+
+        wrapped_->setFieldVL(field, value);
+    }
+
+public:
+    //--------------------------------------------------------------------------
+    // Methods applicable to all ledger entries.
+
+    [[nodiscard]] std::shared_ptr<STLedgerEntry const>
+    slePtr() const
+    {
+        if constexpr (Writable)
+            return std::const_pointer_cast<
+                std::shared_ptr<STLedgerEntry const>>(wrapped_);
+        else
+            return wrapped_;
+    }
+
+    [[nodiscard]] std::shared_ptr<STLedgerEntry> const&
+    slePtr()
+    {
+        static_assert(
+            Writable, "Cannot access non-const SLE of const ledger entry.");
+        return wrapped_;
+    }
```

Comment:
I think it could be problematic to return a `shared_ptr` reference, even `const`, as the pointed to memory could be freed at any time.

Reply:
This interface is provided to provide compatibility with the pre-existing interfaces that are passed `std::shared_ptr<STLedgerEntry> const&`, for example:

- https://github.com/XRPLF/rippled/pull/4280/files#diff-fd246da4e6e9e21f97f1b825c5144977282f5730efad3856f20e11bf4355d1bbR254 and
- https://github.com/XRPLF/rippled/pull/4280/files#diff-fd246da4e6e9e21f97f1b825c5144977282f5730efad3856f20e11bf4355d1bbR201

So unless we we're motivated to do some messing with the existing machinery, I thing we need to return a `std::shared_ptr<STLedgerEntry> const`.

But I do hear your concerns bout returning a reference to a `std::shard_ptr`.  The lifetime of the reference will only be as long as the `LedgerEntryWrapper` that returns the reference.  If we're sufficiently worried about the safely of this interface we could return a copy of the `std::shared_ptr`.  That would provide the lifetime extension properties of a `shared_ptr`, but also incur the cost of copying the `shared_ptr`.

Do you think we should return a copy of the `shared_ptr`?  Is the added safety worth the cost of the atomic increment and decrement in the control block?  Or do you see another alternative that I missed?

Reply:
I honestly don't see why we would pass `std::shared_ptr<STLedgerEntry> const&` to an interface. Passing a raw pointer `STLedgerEntry *` is functionally equivalent, and more versatile in case we happen to have only a raw pointer available.

I also don't like the idea of returning copies of the `std::shared_ptr` because of the cost of doing the memory access to the pointed value to increment (and decrement later) the count, so I would vote for leaving the code as you have it if we can't just use raw pointers when not managing the lifetime of the pointed objects.